### DISCUSSION TITLE
멤버 컨트롤러의 책임 분리

### DISCUSSION THREAD

**Discussion by soeunnPark** [Kernel360/E2E2-CATCHLINE#48] [General]
@sangwonsheep @I-migi

현재 membercontroller에 로그인, 회원 가입, 회원 조회/수정/삭제가 모두 구현되어 있습니다.
API 명세서를 보면 로그인, 회원가입의 경우 /login, /signup
회원 조회/수정/삭제의 경우 /members로 되어 있습니다.

다만 회원 조회/수정/삭제 모두 같은 url로 맵핑되는데 @RequestMapping(/members)로 묶을 수가 없습니다.
그래서 아래와 같이 컨트롤러를 분리하는 것이 어떨까요?

로그인 - LoginController (/login)
회원가입 - SignUpController (/signup)
회원 조회/수정/삭제 - MemberController (/members)
작성한 리뷰 조회/수정/삭제 - MyReviewController
스크랩한 식당 조회 - MyRestaurantController

API 명세서를 보면 위의 기능 모두 MemberController에서 처리하기로 되어 있는데,
컨트롤러도 하나의 책임만 담당하게 하는 것이 좋을 것 같습니다.

어떻게 생각하시는지 남겨주시면 감사할 것 같습니다 😺

**Comment by sangwonsheep** [comment]
저는 로그인/회원가입은 하나의 컨트롤러로 묶어도 좋을 것 같습니다!
반드시 `@RequestMapping`으로 묶을 이유는 없다고 생각합니다.
클래스 레벨에 `@RequestMapping`을 제거하고, 각 메서드에 /login, /logout, /signup 등을 적어주면 좋을 것 같다고 생각합니다!

물론 SRP 원칙에 맞게 하나의 클래스는 하나의 책임을 지도록 쪼개는 것은 맞습니다.
너무 작게 쪼개게 되면 오히려 클래스가 너무 많아지는 문제가 발생할 수 있습니다.
우선은 지금 생각하신대로 최대한 쪼개보고, 음.. 공통으로 묶을 것들이 무엇이 있을까 생각해보시면 좋을 것 같아요!
그래서 쪼갤 때 하나의 컨트롤러에 하나의 메서드만 있기 보다는 추가로 비슷한 역할을 하는 기능끼리 묶으면 좋을 것 같습니다!

그 예시가 제가 말씀 드렸던 로그인/회원가입입니다. 👍 
그 외에 리뷰 조회/수정/삭제로 나누고, 스크랩한 식당 조회로 나눈 부분 칭찬드립니다!! 💯

**Reply by soeunnPark** [reply]
네! 로그인, 로그아웃, 회원가입의 경우 비슷한 역할을 수행하는 것 같습니다. 
또 너무 작게 쪼개게 되면 클래스가 너무 많아져서 비슷한 기능끼리 잘 묶는 것도 좋은 방법인 것 같습니다!

그럼 아래와 같이 나눠서 리팩토링하겠습니다!

로그인 & 로그아웃 & 회원가입 - AuthController (/login, /logout, /signup)
회원 조회/수정/삭제 - MemberController (/members)
작성한 리뷰 조회/수정/삭제 - MyReviewController (/my-page)
스크랩한 식당 조회 - MyRestaurantController (/my-page)

혹시 더 좋은 컨트롤러 이름이나 URL mapping이 있다면 추천해주세요! 😸  
피드백 감사합니다 ~! 😄 😄

**Reply by sangwonsheep** [reply]
컨트롤러 이름, URL mapping 모두 작명 잘하셨습니다!

**Reply by soeunnPark** [reply]
네! 감사합니다 ~! 😄

**Reply by soeunnPark** [reply]
기존에 리뷰와 스크랩이 다음과 같이 나누어져 있었다는 점을 참고해, 아래와 같이 API 명세서 수정했습니다!

<기존>

`/members/{memberId}/reviews`
`/members/{memberId}/scraps`

-> 

<img width="660" alt="스크린샷 2024-08-04 오후 11 20 00" src="https://github.com/user-attachments/assets/d22b78d3-81bd-4f83-acba-2748093fbbb8">
